CIPs/0017.md

```diff
@@ -0,0 +1,69 @@
+
+# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol
+
+- Date: 2020-07-27
+- Author: @lucasege
+- Status: WIP
+
+## Overview
+
+In order to continue supporting the dream of a seamless mobile sync experience, the lightest sync protocol must be extended to support PLUMO Zk-SNARKs which can quickly verify transitions from epoch X to epoch X + Y, where Y could be on the order of 10s-100s of epochs.
+
+The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)
+
+## Goals
+
+- Support adding/storing proofs on full nodes
+- Gossip these proofs to propogate them through the network
+- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.
+
+## Proposed Solution
+
+We propose to add two new RPCs: 
+`AddProof(proof []byte, proofId uint)`
+and
+`Proofs() ([][]byte)`. When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propogate the proof throughout the network.
```

Comment:
How many bytes do we expect the proofs to be?

Reply:
Depends on many factors, but the estimates I've seen for what we may use range from ~5kB - ~12kB. 
The dependent factors include (probably not exhaustive): max number of epochs the circuit is designed for and the type of proof protocol (marlin vs. groth16)

Reply:
We've decided on Groth16 proofs. Marlin and Groth16 both have fully succinct proofs, so the proof size is independent of the number of epochs the circuit is designed for. Groth16 proofs over BW6-761 will be 764B. You will also need to send the client half the instance. The instance includes the public keys at epochs A and B, where B is some number of epochs later. The client will already have validator keys for epoch A and needs the keys at epoch B. 100 public keys over BLS12-377 are 18.9KB.